compiler/src/main/java/com/squareup/anvil/compiler/IrUtils.kt

```diff
@@ -0,0 +1,105 @@
+package com.squareup.anvil.compiler
+
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.ir.declarations.IrAnnotationContainer
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationWithName
+import org.jetbrains.kotlin.ir.declarations.IrValueParameter
+import org.jetbrains.kotlin.ir.expressions.IrConstructorCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrMemberAccessExpression
+import org.jetbrains.kotlin.ir.expressions.IrVararg
+import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
+import org.jetbrains.kotlin.ir.symbols.IrClassifierSymbol
+import org.jetbrains.kotlin.ir.types.IrSimpleType
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classifierOrNull
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.typeOrNull
+import org.jetbrains.kotlin.ir.util.fqNameWhenAvailable
+import org.jetbrains.kotlin.ir.util.getAnnotation
+import org.jetbrains.kotlin.ir.util.getArgumentsWithIr
+import org.jetbrains.kotlin.name.FqName
+
+internal fun IrPluginContext.requireReferenceClass(fqName: FqName): IrClassSymbol {
+  return referenceClass(fqName) ?: throw AnvilCompilationException(
+      message = "Couldn't resolve reference for $fqName"
+  )
+}
+
+internal fun IrAnnotationContainer.annotationOrNull(
+  annotationFqName: FqName,
+  scope: FqName? = null
+): IrConstructorCall? {
+  val annotation = getAnnotation(annotationFqName)
```

Comment:
I think this whole thing could just be:

```kotlin
return getAnnotation(annotationFqName)?.takeIf { annotation ->
  scope != null && annotation.scope() == scope
}
```

Reply:
It's an OR, but nonetheless you're right!
```
  return getAnnotation(annotationFqName)?.takeIf { annotation ->
    scope == null || annotation.scope() == scope
  }
```